### ISSUE TITLE
Exception subscriptions?

### ISSUE THREAD

**Issue by michi42** [ossgang/ossgang-commons#7]
should we in general have something like
```java
interface Observable<T> {
   Subscription subscribe(Consumer<T>, Options ...)
   Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
}
```
(with the explicit premise that a delivered exception does NOT kill any subscription or collapse anything)

This could be both exceptions from upstream (like a JAPC exception if a device is down) and exception that occur during transformations.

The alternative is of course an ``Observable<Maybe<T>>`` ;-)

**Comment by michi42** [comment]
(Originally posted by @kaifox in !2)
> What I meant - should we in general have something like
> 
> ```java
> interface Observable<T> {
>    Subscription subscribe(Consumer<T>, Options ...)
>    Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
> }
Mhmm ... if we do it, I would tend to prefer the `subscribeException` version over the `Observable<Maybe<T>>`...

Another option could be (without thinking too much if it would really work)

```
interface Observable<T> {
   Subscription subscribe(Consumer<T>, Options ...);
   Observable<Throwable> exceptions();
   Observable<Maybe<T>> asMaybes();
}
```

... and I definitely would avoid collapsing things ;-) (Except if we introduce another option to do so ;-)

**Comment by michi42** [comment]
> Observable<Throwable> exceptions();

But then you could have exceptions of exceptions of exceptions ;-)
but I like the asMaybe()

**Comment by michi42** [comment]
(originally posted by @kaifox in #5)

> > Observable exceptions();
> > But then you could have exceptions of exceptions of exceptions ;-)
> > but I like the asMaybe()

I agree ... I thought the same ...

Could another interface E.g. `ErrorDeflectingObservable` help which would have the two additional methods? Probably really not pollute the ObservableInterface ...

Probably it is one of the candidates to leave out for later ;-)

**Comment by michi42** [comment]
I could even imagine something like this:
```java
interface Observable<T> {
   Subscription subscribe(Consumer<Maybe<T>>, Options ...)
   Subscription subscribeValues(Consumer<T>, Options ...)
   Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
}
```
but that may go a bit to far ;-)

I think the basic question is if we want to build into the infrastructure a way to propagate (and possibly even handle) errors from the source, or if we say they would anyway in most cases "just" get logged, we don't complicate the interfaces for that, and in case it is needed one should explicitly create an Observable<Maybe<...>>.

**Comment by kaifox** [comment]
Hi again, (It seems I miss updates recently ;-)

My opinion:
* I think the `subscribe()` method should be on the pure `T`,
* The others should be qualified (Exceptions, Maybes)...
* Ideally, there would be a way to propagate errors downstream... (e.g. through the maybe channel)

... to be seen if this does not get too complicated...

Working hypothesis:
```
interface Observable<T> {
   Subscription subscribe(Consumer<T>, Options ...)
   Subscription subscribeMaybes(Consumer<Maybe<T>, Options ...)
   Subscription subscribeExceptions(Consumer<Throwable>, Options ...)
}
```
(btw the plural of maybe is akward ;-)

... Would it be more concise/natural to have something like:

```
interface Observable<T> {
   Subscription onValue(Consumer<T>, Options ...)
   Subscription onMaybe(Consumer<Maybe<T>, Options ...)
   Subscription onException(Consumer<Throwable>, Options ...)
}
```

?

**Comment by kaifox** [comment]
And: Yes, the 'deflect' might not be a good name...

**Comment by kaifox** [comment]
I think I like 'resilient' .... it expresses nicely that the thing does not colapse ;-)

**Comment by kaifox** [comment]
Thinking a bit more (and after writing my comment on slack ;-), I would even brainstorm-wise introduce another observable derivative.... 

```
interface MaybeObservable<T> extends Observable<Maybe<T>> {
// tbd
}
```

Following my recent proposal, then `ResilientObservable` would potentially look like this:

```
interface ResilientObservable<T> extends Observable<T> {
   MaybeObservable<T> maybes();
   Observable<Throwable> throwables(); // or exceptions() ?

  default Subscription onMaybe(Consumer<Maybe<T>> c, Options ... o) {
    return maybes().onValue(c, o);
  }

  default Subscription onThrowable(Consumer<Throwable> t, Options ... o) {
    return throwables().onValue(t, o);
  }
}
```

... the advantage being, that e.g. `MaybeObservable` would have mapping methods like:
```
interface MaybeObservable<T> extends Observable<Maybe<T>> {
  <U> MaybeObservable<U> map(Function<T,U> mapper);
}
```

Instead of 
```
  <U> Observable<U> map(Function<Maybe<T>,Maybe<U>> mapper);
```
as it would come from the Observable ...

(to be tried if this really works out)

**Comment by michi42** [comment]
Solved by #10.
